### DISCUSSION TITLE
[Precomputed ControlNet] Speed up ControlNet by 45% - but is it necessary?

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/ControlNet#216] [Announcements]
Hi everyone, we plan to experiment with a feature called "Precomputed ControlNet". This can be achieved by modifying 2 or 3 lines of the training code to progressively disconnect the input concat here:

![image](https://user-images.githubusercontent.com/19834515/222873546-b8b13190-8925-4f5e-bdb3-9d45c0a521e4.png)

By doing this, we will be able to execute controlnet only one time before diffusion, rather than in every diffusion iteration. The controlnet should be [equally powerful and robust as before](https://github.com/lllyasviel/ControlNet/discussions/188). This will lead to a speed up by about 40% to 45%. And this will further decrease the required GPU memory.

Nevertheless, if we can observe any performance decrease (even minimal decrease) in any experimental setting (including no prompt setting, short prompt setting, and long prompt setting), we will make this experimental feature on-hold to avoid confusing and to prevent giving new users a mis-estimate of the models' capabilities.

Let us know what you think about it! Thank you for your support as always.

**Update (after all experiments of cn1.1): this experiment fails to train ControlNets as good as the proposed implementation in our paper. We observe that models trained with this method tends to produce more artifacts and less robust results. We have given up this feature. The input from each diffusion steps is necessary for robustness, and necessary for special models like Shuffle and IP2P (in controlnet 1.1).**

**Comment by yakuzadave** [comment]
Great work so far.  This has been a lot of fun to play with.

**Comment by axsddlr** [comment]
I am down to try it out

**Comment by OedoSoldier** [comment]
Good idea, afaik many ppl w/ low vram GPUs are struggling w/ using CN w/ highres fix so if it works it'll be great.

**Comment by TheLukaDragar** [comment]
Great which 3 lines did you change? I would like to try it.

**Comment by drbobo0** [comment]
looks awesome , keep it up

**Comment by Njasa2k** [comment]
Is this speedup competitive with T2I adapter?

**Comment by Aridea2021** [comment]
that's good news for me, 期待早点体验，目前gpu vram总是不够，计算也很慢

**Comment by neverix** [comment]
Wouldn't this let you use a much larger model

**Comment by Aridea2021** [comment]
autoreply：Your mail has been received, and I will reply as soon as I canBest Regards；）

**Comment by dan4ik94** [comment]
Yes, it's really necessary. I have a 4GB GPU and I would benefit of this. Thank you for you hard work.

**Comment by xueqing0622** [comment]
Good idea,  I need it very much,

**Comment by Georgefwt** [comment]
Could we use the released model to achieve this or should we retrain a new model to achieve this?
I tried to tweak the sampling steps in the way you described and used the released model, and it didn't work.😢

**Comment by ffdown** [comment]
we need it =))

**Comment by lioo717** [comment]
how to deal with the timesteps? Does it use T at the first step, and the T result for all following steps?

**Comment by Aridea2021** [comment]
autoreply：Your mail has been received, and I will reply as soon as I canBest Regards；）

**Comment by lllyasviel** [comment]
Update: this experiment fails to train ControlNets as good as the standard implementation. We have given up this feature. The input from each diffusion steps is necessary for robustness, and necessary for special models like Shuffle and IP2P.

**Reply by Njasa2k** [reply]
can we see the images?

**Comment by Aridea2021** [comment]
autoreply：Your mail has been received, and I will reply as soon as I canBest Regards；）

**Comment by karkra911** [comment]
can we use  NAS (neural architecture search) for even more GPU efficiency ? 

 finding new neural architecture which is even much faster ?

does anyone have Nvidia dgx workstation to run NAS , and find a better and even more gpu efficient and fast algorithm ?

**Comment by Aridea2021** [comment]
autoreply：Your mail has been received, and I will reply as soon as I canBest Regards；）
